lib/logical.js

```diff
@@ -89,44 +94,54 @@ function insertLeaf(tree, leaf) {
   entry[leaf.name] = leaf;
 }
 
-function problem(root, issue) {
-  if (!root.problems) {
-    root.problems = [];
+function walkDeps(root, tree, from) {
+  if (!from) {
+    from = tree.__from;
   }
-  root.problems.push(issue);
-}
 
-function walkDeps(root, tree) {
   // only include the devDeps on the root level package
   var deps = _.extend({}, tree.__dependencies,
-    tree.__from && tree.__from.length === 1 ? tree.__devDependencies : {});
-  return Object.keys(deps).reduce(function walkDepsPicker(acc, curr) {
-    var version = deps[curr];
-    var dep = pluck(root, tree.__from, curr, version);
+    tree.__from && from.length === 1 ? tree.__devDependencies : {});
 
-    if (!dep) {
-      problem(root, 'missing: ' + curr + '@' + version +
-              ', required by ' + tree.name + '@' + tree.version);
-      return acc;
-    }
+  return Object.keys(deps).reduce(function walkDepsPicker(acc, curr) {
+    // only attempt to walk this dep if it's not in our path already
+    if (tree.__from.indexOf(curr) === -1) {
```

Comment:
Key change - only walk these deps if we don't have them in the tree already (preventing infinite depth).